ice40/utils/fasm_icebox/fasm_icebox_utils.py

```diff
@@ -0,0 +1,596 @@
+        if feature.parts[-1] == "SimpleInput":
+            iob = int(feature.parts[-2][-1])
+            new_ieren = find_ieren(ic, feature.loc, iob)
+
+            feature.parts[-1] = "PINTYPE_0"
+            db_entry = fasmdb[feature.toFasmEntry().feature]
+            _set_feature_bits(ic, db_entry.loc, db_entry.bit_tuples)
+
+            feature.loc = new_ieren[:2]
+            feature.parts[-2] = "IoCtrl"
+            feature.parts[-1] = "IE_{}".format(iob)
+            db_entry = fasmdb[feature.toFasmEntry().feature]
+            _set_feature_bits(ic, db_entry.loc, db_entry.bit_tuples)
+            feature.parts[-1] = "REN_{}".format(iob)
+            db_entry = fasmdb[feature.toFasmEntry().feature]
+            _set_feature_bits(ic, db_entry.loc, db_entry.bit_tuples)
+            continue
+
+        if feature.parts[-1] == "SimpleOutput":
+            iob = int(feature.parts[-2][-1])
+            new_ieren = find_ieren(ic, feature.loc, iob)
+
+            feature.parts[-1] = "PINTYPE_3"
+            db_entry = fasmdb[feature.toFasmEntry().feature]
+            _set_feature_bits(ic, db_entry.loc, db_entry.bit_tuples)
+            feature.parts[-1] = "PINTYPE_4"
+            db_entry = fasmdb[feature.toFasmEntry().feature]
+            _set_feature_bits(ic, db_entry.loc, db_entry.bit_tuples)
+
+            feature.loc = new_ieren[:2]
+            feature.parts[-2] = "IoCtrl"
+            feature.parts[-1] = "REN_{}".format(iob)
+            db_entry = fasmdb[feature.toFasmEntry().feature]
+            _set_feature_bits(ic, db_entry.loc, db_entry.bit_tuples)
+            continue
+
+        ## special case for RAM INIT values
+        tile_type, loc = feature.tile_type, feature.loc
+        if tile_type == "RAMB" and feature.parts[-1].startswith("INIT"):
+            tloc = tuple(loc)
+            if tloc not in ic.ram_data:
+                ic.ram_data[tloc] = [64 * "0" for _ in range(16)]
+            tile = ic.ram_data[tloc]
+            tile_bits = _tile_to_array(tile, is_hex=True)
+        elif tile_type == "RAMB" and feature.parts[-1].endswith("_MODE"):
+            # hack to force modes to the RAMT
+            tile = ic.tile(loc[0], loc[1] + 1)
+            tile_bits = _tile_to_array(tile)
+        else:
+            tile = ic.tile(*loc)
+            tile_bits = _tile_to_array(tile)
+
+        # lookup feature and convert
+        for canonical_feature in fasm.canonical_features(line.set_feature):
+            key = fasm.set_feature_to_str(canonical_feature)
+            feature = fasmdb[key]
+            bm, bv = _get_feature_bits(tile_bits, feature.bit_tuples)
+
+            tile_bits[bm] = bv[bm]
+
+        if tile_type == "RAMB" and feature.parts[-1].startswith("INIT"):
+            _array_to_tile(tile_bits, tile, is_hex=True)
+        else:
+            _array_to_tile(tile_bits, tile)
+
+    # TODO: would be nice to upstream a way to write to non-files
+    ic.write_file(outf.name)
+
+
+class TestConversion(unittest.TestCase):
```

Comment:
Should the test case be in the file itself?

Reply:
Not sure if we have a convention for how we want to do this, but I did move it to a separate file similar to rr_graph